org.opent2t.sample.lamp.superpopular/com.hue.bulb/js/colour.js

```diff
@@ -0,0 +1,382 @@
+"use strict";
+
+const ChangeTolerance = 0.0001;
+const MaxHue = 360.0;
+const MaxColour = 255;
+
+/*
+ * Method to see if the given XY value is within the reach of the lamps.
+ * 
+ * Details on the math theory is avaialble at http://mathworld.wolfram.com/TriangleInterior.html
+ * and https://en.wikipedia.org/wiki/Barycentric_coordinate_system
+ *
+ * @param xyColour is a two-element object representing a point containing the X,Y value
+ * @param colourGamut   the color gamut of the light device.
+ * @return true if within reach, false otherwise.
+ */
+function isPointInLampsReach(xyColour, colourGamut){
+
+    //Calculate vectors based on colourGamut[0]
+    var v1 = [colourGamut[1].x - colourGamut[0].x, colourGamut[1].y - colourGamut[0].y];
+    var v2 = [colourGamut[2].x - colourGamut[0].x, colourGamut[2].y - colourGamut[0].y];
+    var v = [xyColour.x - colourGamut[0].x, xyColour.y - colourGamut[0].y];
+
+    //Calculate the barycentric coordinates
+    var a = determinant(v, v2) / determinant(v1, v2);
+    var b = determinant(v1, v) / determinant(v1, v2);
+
+    return ((a >= 0) && (b >= 0) && (a + b <= 1)) ? true : false;
+}
+
+function determinant(pt1, pt2) {
+    return ((pt1.x * pt2.y) - (pt1.y * pt2.x));
+}
+
+/*
+ * Find the closest point on a line.
+ * This point will be within reach of the lamp.
+ *
+ * Details on the math theory is avaiable at 
+ * https://en.wikipedia.org/wiki/Distance_from_a_point_to_a_line#A_vector_projection_proof
+ *
+ * @param ptA the point where the line starts
+ * @param ptB the point where the line ends
+ * @param ptP the point which is close to a line.
+ * @return the point which is on the line.
+ */
+function getClosestPointToPoints(ptA, ptB, ptP){
+
+    var AP = {x: (ptP.x - ptA.x), y: (ptP.y - ptA.y)};      // Vector from ptA to ptP
+    var AB = { x: (ptB.x - ptA.x), y: (ptB.y - ptA.y) };    // Vector from ptA to ptB
```

Comment:
Fixed.

---
In reply to: [94821900](https://github.com/openT2T/translators/pull/143#discussion_r94821900) [](ancestors = 94821900)